overlord/udevmon.go

```diff
@@ -85,16 +87,59 @@ func (m *UDevMonitor) Connect() error {
 // handles hotplug events (devices added or removed). It returns immediately.
 // The goroutine must be stopped by calling Stop() method.
 func (m *UDevMonitor) Run() error {
+	existingDevices := make(chan *hotplug.HotplugDeviceInfo)
+	udevadmErrors := make(chan error)
+
+	if err := hotplug.EnumerateExistingDevices(existingDevices, udevadmErrors); err != nil {
+		return fmt.Errorf("cannot enumerate existing devices: %s", err)
+	}
+
 	m.tmb.Go(func() error {
+		var enumerationFinished bool
+		// Gather devices from udevadm info output (enumeration on startup) and from udev event monitor:
+		// - devices discovered on startup are reported via existingDevices channel
+		// - added/removed devices are reported via netlinkEvents channel
+		// Devices reported by netlinkEvents channel are queued until all devices from existingDevices
+		// are processed.
+		// It might happen that a device is plugged at startup and is reported by both existingDevices
+		// channel and netlinkEvents channel; such events are ignored by de-duplication logic based on device path;
+		// this de-dup logic is only applied until enumeration is finished.
 		for {
 			select {
+			case err, ok := <-udevadmErrors:
+				if ok {
+					logger.Noticef("udevadm error: %q\n", err)
+				} else {
+					udevadmErrors = nil
+				}
+			case dev, ok := <-existingDevices:
+				if ok && m.deviceAddedCb != nil {
+					m.deviceAddedCb(dev)
+					m.startupDevices[dev.DevicePath()] = true
+				}
+				if !ok {
+					existingDevices = nil
+					// enumeration of existing devices has finished, flush queued events
+					for _, ev := range m.queuedNetlinkEvents {
+						m.udevEvent(ev)
+					}
+					m.queuedNetlinkEvents = nil
+					m.startupDevices = nil
+					enumerationFinished = true
+				}
 			case err := <-m.netlinkErrors:
 				logger.Noticef("netlink error: %q\n", err)
 			case ev := <-m.netlinkEvents:
-				m.udevEvent(&ev)
+				// queue netlink events until enumeration of existing devices finishes
+				if !enumerationFinished {
+					m.queuedNetlinkEvents = append(m.queuedNetlinkEvents, &ev)
```

Comment:
That sounds unnecessarily tricky. If we want them in sequence, why not doing exactly that and not attempting to consume netlink events while we're still working on the enumeration? So enumerate first, generate all necessary events, then listen to netlink events. Many of the details here will go away I think.